osu.Game/Online/Chat/ChannelManager.cs

```diff
@@ -0,0 +1,298 @@
+        }
+
+        private void currentChannelChanged(Channel channel)
+        {
+            if (!JoinedChannels.Contains(channel))
+                JoinedChannels.Add(channel);
+        }
+
+        /// <summary>
+        /// Posts a message to the currently opened channel.
+        /// </summary>
+        /// <param name="text">The message text that is going to be posted</param>
+        /// <param name="isAction">Is true if the message is an action, e.g.: user is currently eating </param>
+        public void PostMessage(string text, bool isAction = false)
+        {
+            if (CurrentChannel.Value == null)
+                return;
+
+            if (!api.IsLoggedIn)
+            {
+                CurrentChannel.Value.AddNewMessages(new ErrorMessage("Please sign in to participate in chat!"));
+                return;
+            }
+
+            var message = new LocalEchoMessage
+            {
+                Sender = api.LocalUser.Value,
+                Timestamp = DateTimeOffset.Now,
+                TargetType = CurrentChannel.Value.Target,
+                TargetId = CurrentChannel.Value.Id,
+                IsAction = isAction,
+                Content = text
+            };
+
+            CurrentChannel.Value.AddLocalEcho(message);
+
+            var req = new PostMessageRequest(message);
+            req.Failure += e => CurrentChannel.Value?.ReplaceMessage(message, null);
+            req.Success += m => CurrentChannel.Value?.ReplaceMessage(message, m);
+            api.Queue(req);
+        }
+
+        public void PostCommand(string text)
+        {
+            if (CurrentChannel.Value == null)
+                return;
+
+            var parameters = text.Split(new[] { ' ' }, 2);
+            string command = parameters[0];
+            string content = parameters.Length == 2 ? parameters[1] : string.Empty;
+
+            switch (command)
+            {
+                case "me":
+                    if (string.IsNullOrWhiteSpace(content))
+                    {
+                        CurrentChannel.Value.AddNewMessages(new ErrorMessage("Usage: /me [action]"));
+                        break;
+                    }
+
+                    PostMessage(content, true);
+                    break;
+
+                case "help":
+                    CurrentChannel.Value.AddNewMessages(new InfoMessage("Supported commands: /help, /me [action]"));
+                    break;
+
+                default:
+                    CurrentChannel.Value.AddNewMessages(new ErrorMessage($@"""/{command}"" is not supported! For a list of supported commands see /help"));
+                    break;
+            }
+        }
+
+        private void fetchNewMessages()
+        {
+            if (fetchChannelMsgReq == null)
+                fetchNewChannelMessages();
+
+            if (fetchUserMsgReq == null)
+                fetchNewUserMessages();
+        }
+
+        private void fetchNewUserMessages()

```

Comment:
looks very similar to fetchNewChannelMessages, you can try to create fetchMessages(APIRequest r)

Reply:
? Please explain

Reply:
```
fetchMessages(APIRequest requset, Action<IEnumerable<Message>> handler) { code from fetchNewUserMessages }

fetchMessages(new GetUserMessagesRequest(lastUserMsgId), handleUserMessages)
fetchMessages(new GetChannelMessagesRequest(JoinedChannels.Where(c => c.Target == TargetType.Channel), lastChannelMsgId)), handleChannelMessages)
```